### DISCUSSION TITLE
How to collect info from your RV to make a floor-plan

### DISCUSSION THREAD

**Discussion by spbrogan** [spbrogan/rvc2mqtt#22] [General]
Currently,

Any RVC message that isn't handled by an entity instance will go to a special logger. 
That logger can be setup to log to a file.
Then user can trigger things in RV and review the log file.

This needs access to the host and is not user friendly.  
Better ideas?

**Comment by AllonisDavid** [comment]
Is your question, how can you control your RV from a floorplan view?  This is what our myServer 6 application is perfectly suited for.  https://allonis.com  We would just have to write a driver to subscribe to the MQTT RVC messages.  And leave the rest to myServer.  If you would like to do this, let me know.

**Reply by spbrogan** [reply]
Not really about the "view".   If you look halfway thru the `configuration.yaml` file (https://github.com/spbrogan/rvc2mqtt/blob/main/docs/configuration.md)  you will see a `map` section.  
This section basically maps the RV-C device nodes to actual things like "lights", front-ac, rear-ac, etc.  

Basically, RVs have different nodes and different devices, and some vendors use different devices to map different things.  I have heard of dimmer devices that actually control awnings and things like that.  Does your "myServer" provide a way for the user to figure out that configuration?  Do you have to maintain a list of these for your users?  I assume there could be at least one map per model of rv per year (or more).

**Reply by AllonisDavid** [reply]
Frequently in "home automation", dimmer levels are used for blinds as they are light control devices.  Awnings use same motors.  Values of 0 - 100 are the most common values.
So, we have graphical sliders where the user can go "50%" and the blind / awning will go out to the 50% mark (or the closest preset).
The user interface is what "figures out" what the device actually is ie: on the floorplan, you can click on the awning and control that device.  The graphical object is "mapped" to the device that is the awning motor.

You would not organize this by model / model year.  That becomes an impossible to manage issue.  Only the particular system needs to know which devices are what type of device.   The end user / integrator can create an Alias to say "Lighting device # 232 is "Main Awning".  The system only needs to know "user wants to set device 232 to 50% position.  The Control system then sends a MQTT message to your software, that then sends the right RV-C message on the CAN bus.  The Awning motor then "hears" that message and goes to the 50% position (if it supports partial positions).

Keep the RV-C MQTT simple...a way to get the status value to the control system ("myServer" is the control system).  And for the control system to send commands to RV-C network devices that they support.  I would not define device names in your config file.  Then the user has to understand that they need to do that and how.  This is the role of the control system, not the device interface.

In myServer, we have web templates that is the user interface definition.  This is where the RV's floorplan would be defined.  The awning would be a graphical object next to the RV.  The myServer database then keeps the sofware Drivers used in the project (ie: the "RV-C_MQTT" driver), the devices that driver reports back to exist (the Awning), the macros for the system (a sequence of instructions like "bring the main awning to 0 (In) and then the window awning to 0).  And the automation events "when the wind goes over 10mph for more than 2 minutes, trigger the Awning In macro".  The collective of these elements defines the system for THAT RV and Those Users and their preferences and services.

Here is an example of a Lighting driver (Z-Wave technology).  The "Device Address" is what the RV-C_MQTT interface would provide in the MQTT JSON.  The "ID" is how myServer tracks that device across all devices.  The Model would be information the RV-C_MQTT interface could provide.  The "Family" is what types of commands the device supports.  The Awning device would be associated to the Lighting Family.    The RV-C_MQTT interface would support multiple Families of devices (Climate, Security, OBD, Energy, etc). The "Name" is the user typed Alias that they then know what that device is "Main Awning".

![image](https://user-images.githubusercontent.com/63688643/150920496-0a272d2b-9108-4a60-8e28-692366b1e7fb.png)

**Comment by spbrogan** [comment]
for now this has been done using two yaml files.  This creates devices and entities in HA.  
For those wanting a pure RV-C to MQTT bridge that just opens up communication to rv-c that will have to be a different project.  I think it would be pretty easy...but this project is creating entities and devices and working with HA integrations using MQTT.  

Thanks

**Comment by AllonisDavid** [comment]
Hey Sean,
Would like to chat with you about the RV-C MQTT project.
Can you give me a call?
David
Allonis LLC
248-918-0123

**Reply by rlust** [reply]
Any movement here on this mapping function? I am interested in giving it a try!

**Reply by AllonisDavid** [reply]
Well, depends on what definition of "mapping" you are speaking of.
From my perspective, we just started doing the work to integrate Sean's RV-C MQTT project into our myServer 6 control system.
We have OBDII working too.  J1939 (diesel) has just started.  RV-C is J1939 as well.
We are getting RV-C MQTT going with HomeAssistant first as Sean originally wrote it for that purpose.  Allonis will then determine best strategy to migrate it for myServer 6 use.  BTW, we also have most of a HomeAssistant to myServer 6 API integration complete as well.

**Reply by spbrogan** [reply]
> Any movement here on this mapping function? I am interested in giving it a try!

I haven't made any changes in months.  It has been extremely reliable in my RV and works 100x better than the OnCommand software/app provided with the RV.  The mapping function/process, as provided in this solution, is you install the software with an empty floor plan configuration, and configure logging as shown in the example: https://github.com/spbrogan/rvc2mqtt/blob/main/docs/configuration.md#example-1 then you just need to trigger devices in the RV and watch the unhandled log output.  From there you can easily build up the "floorplan" which maps RV-C commands to "entities".  These will then show up in mqtt and in HA if you have mqtt autodetect turned on.   Let me know if i can help.
